File: personlighetstyper/personlighetstyper_1.cpp
#include <iostream>
#include <vector>

int main() {
  std::cin.tie(nullptr);
  std::ios_base::sync_with_stdio(false);

  int n, m;
  std::cin >> n >> m;

  std::string name;
  std::cin >> name;

  int32_t forbidden[26];
  for (int i = 0; i < 26; i++) {
    forbidden[i] = 0;
  }

  for (int i = 0; i < m; i++) {
    char a, b;
    std::cin >> a >> b;
    
    forbidden[b - 'A'] |= 1 << (a - 'A');
  }

  std::vector<int64_t> dp(n + 1, 1000000000ll);
  dp[0] = 0;

  for (int i = 0; i < n; i++) {
    int64_t x = dp[i];

    int32_t here = 0;
    int32_t incompatible = 0;
    for (int j = i + 1; j <= n; j++) {
      char ch = name[j - 1];

      here |= 1 << (int32_t)(name[j - 1] - 'A');
      incompatible |= forbidden[name[j - 1] - 'A'];

      if (~incompatible & here) {
        dp[j] = std::min(dp[j], x + 1);
      }
    }
  }

  std::cout << dp[n];
}

File: personlighetstyper/personlighetstyper_1.py
n, m = map(int, input().split())

name = input()

forbidden = dict()
for ch in range(ord('A'), ord('Z') + 1):
    forbidden[chr(ch)] = set()

for _ in range(m):
    a, b = input().split()

    forbidden[b].add(a)

dp = [100_000_000] * (n + 1)
dp[0] = 0

for i in range(n):
    x = dp[i]

    here = set()
    incompatible = set()

    for j in range(i + 1, n + 1):
        here.add(name[j - 1])
        for ch in forbidden[name[j - 1]]:
            incompatible.add(ch)

        for ch in here:
            if ch not in incompatible:
                dp[j] = min(dp[j], x + 1)
                break

print(dp[n])
